ca_tools/scripts/pid_controller.py

```diff
@@ -0,0 +1,245 @@
+        self._my_vel_pub = rospy.Publisher(VEL_PUB_TOPIC, Twist, queue_size=10)
+        self._my_pid_lineal_state_pub = rospy.Publisher(
+            "/lineal/state", Float64, queue_size=1)
+        self._my_pid_angular_effort_sub = rospy.Subscriber(
+            "/lineal/control_effort", Float64, self._lineal_effort_callback)
+        self._my_pid_lineal_setpoint_pub = rospy.Publisher(
+            "/lineal/setpoint", Float64, queue_size=10)
+        self._my_gts_sub = rospy.Subscriber(
+            GTS_SUB_TOPIC, Odometry, self._callback)
+        self._my_pid_angular_effort_sub = rospy.Subscriber(
+            "/angular/control_effort", Float64, self._angular_effort_callback)
+        self._my_path_publisher = rospy.Publisher(
+            "/create1/my_path", Path, queue_size=10)
+        self._my_path = Path()
+        self._my_pid_angular_setpoint_pub = rospy.Publisher(
+            "/angular/setpoint", Float64, queue_size=1)
+
+        self._path_threshold = PATH_THRESHOLD  # Throttle the path pub
+        self._my_goals = [Pose2D()] * 4
+        self._my_goals[1] = Pose2D(0, 2, 0)
+        self._my_goals[2] = Pose2D(2, 2, 0)
+        self._my_goals[3] = Pose2D(2, 0, 0)
+
+        self._lineal_state = None
+        self._goal_num = 0
+        self._state = STATES.STOP_MOVING_FORWARD
+        self._my_pose = Pose2D()
+        self._my_pid_state = 0.0
+        self._my_angle_goal = None
+        self._my_pid_angular_effort = 0.0
+        self._my_pid_lineal_effort = 0.0
+
+    def _angular_effort_callback(self, data):
+        """Callback for the angular effort calculated by the PID node
+
+        Arguments:
+            data {[Float64]} -- [Effort]
+        """
+        self._my_pid_angular_effort = data.data
+
+    def _lineal_effort_callback(self, data):
+        """Callback for the angular effort calculated by the PID node
+
+        Arguments:
+            data {[Float64]} -- [Effort]
+        """
+        self._my_pid_lineal_effort = data.data
+
+    def _callback(self, data):
+        """Callback for groundtruth
+
+        Arguments:
+            data {[Odometry]} -- [Odometry msg from gts]
+        """
+        self._my_pose.x = data.pose.pose.position.x  # Update pose
+        self._my_pose.y = data.pose.pose.position.y
+        q = data.pose.pose.orientation  # get orientation quaternion
+        self._path_threshold -= 1
+        q_arr = [q.x, q.y, q.z, q.w]
+        euler = euler_from_quaternion(q_arr)
+        self._my_pose.theta = euler[2]  # theta=yaw
+        if (self._path_threshold == 0):  # add the pose to path every PATH_THRESHOLD cycles
+            aux = PoseStamped()
+            aux.pose = data.pose.pose
+            aux.header.frame_id = "/map"
+            self._my_path.header.frame_id = "/map"
+            self._my_path.poses.append(aux)
+            self._my_path_publisher.publish(self._my_path)
+            self._path_threshold = PATH_THRESHOLD
+
+    def _move(self):
```

Comment:
Is there a way of splitting the code to don't duplicate methods like this and in case you use the PID package of your own implementation, you can use the same logic.

Reply:
You are talking about the `move()` method? Maybe you can give me a hint? I don't think i'm understanding what you say. Maybe i could have a class with a `move()` method and herit from it?

Reply:
You have two methods called `_move` throughout the code and they do the same. The same for other methods.

Reply:
Changed `PidCtrlNode` to herit from `CtrlNode` in order to reuse some code, tell me if that's enough or you can think of another way of improving it.